src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithAllFeatures.cs

```diff
@@ -99,14 +81,94 @@ public void It_publishes_the_project_correctly(string targetFramework, string []
 }");
             baselineConfigJsonObject["runtimeOptions"]["tfm"] = targetFramework;
-            baselineConfigJsonObject["runtimeOptions"]["framework"]["version"] = 
+            baselineConfigJsonObject["runtimeOptions"]["framework"]["version"] =
                 targetFramework == "netcoreapp1.0" ? "1.0.5" : "1.1.2";
 
             runtimeConfigJsonObject
                 .Should()
                 .BeEquivalentTo(baselineConfigJsonObject);
         }
 
+        [Fact]
+        public void It_fails_when_nobuild_is_set_and_build_was_not_performed_previously()
+        {
+            var publishCommand = GetPublishCommand("netcoreapp1.0").Execute("/p:NoBuild=true");
+            publishCommand.Should().Fail().And.HaveStdOutContaining("MSB3030"); // "Could not copy ___ because it was not found."
+        }
+
+        [Theory]
+        [MemberData(nameof(PublishData))]
+        public void It_does_not_build_when_nobuild_is_set(string targetFramework, string[] expectedPublishFiles)
+        {
+            var publishCommand = GetPublishCommand(targetFramework);
+
+            // do a separate build invocation before publish
+            var buildCommand = new BuildCommand(Log, publishCommand.ProjectRootPath);
+            buildCommand.Execute().Should().Pass();
+
+            // modify all project files, which would force recompilation if we were to build during publish
+            foreach (string projectFile in EnumerateFiles(buildCommand, "*.csproj"))
+            {
+                File.AppendAllText(projectFile, " ");
+            }
+
+            // capture modification time of all binaries before publish
+            var modificationTimes = GetLastWriteTimesUtc(buildCommand, "*.exe", "*.dll", "*.resources", "*.pdb");
+
+            // publish (with NoBuild set)
+            publishCommand.Execute("/p:NoBuild=true").Should().Pass();
+            publishCommand.GetOutputDirectory(targetFramework).Should().OnlyHaveFiles(expectedPublishFiles);
+
+            // check that publish did not modify any of the build output
+            foreach (var (file, modificationTime) in modificationTimes)
+            {
+                File.GetLastWriteTimeUtc(file)
+                    .Should().Be(
+                        modificationTime, 
+                        because: $"Publish with NoBuild=true should not overwrite {file}");
+            }
+        }
+
+        private static List<(string, DateTime)> GetLastWriteTimesUtc(MSBuildCommand command, params string[] searchPatterns)
+        {
+            return EnumerateFiles(command, searchPatterns)
+                .Select(file => (file, File.GetLastWriteTimeUtc(file)))
+                .ToList();
+        }
+
+        private static IEnumerable<string> EnumerateFiles(MSBuildCommand command, params string[] searchPatterns)
+        {
+            return searchPatterns.SelectMany(
+                pattern => Directory.EnumerateFiles(
+                    Path.Combine(command.ProjectRootPath, ".."),
```

Comment:
I'm guessing this is going up one level from the `ProjectRootPath` in order to cover the referenced projects, but that's worth an explanatory comment.

Reply:
Correct. I will add a comment.

Reply:
Fixed.